src/allmydata/storage_client_http.py

```diff
@@ -0,0 +1,204 @@
+
+    Some day ``IStorageServer`` might go away, but for now this is the proposed
+    strategy for support the new HTTP protocol.
+
+    NOTE: For now this is just a sketch, to demonstrate this approach is
+    viable.  The other part of this approach is to rename
+    ``IFoolscapStorageServer`` to something more generic, and try to paper over
+    the differences between Foolscap and HTTP there (e.g. use X.509 cert hash
+    instead of tub ID), but ``IStorageServer`` is more fundamental so this is
+    where we're starting.
+    """
+    def __init__(self, storage_client):
+        # type: (IStorageClientV2) -> None
+        self._client = storage_client
+
+    def get_version(self):
+        return self._client.get_version()
+
+    @inlineCallbacks
+    def allocate_buckets(
+            self, storage_index, renew_secret, cancel_secret, sharenums,
+            allocated_size, canary
+    ):
+        result = yield self._client.immutable_create(
+            storage_index, sharenums, allocated_size, renew_secret,
+            cancel_secret
+        )
+        returnValue(
+            result.already_got,
+            {
+                share_num: _ClientV2BucketWriter(
+                    self._client, storage_index, share_num
+                )
+                for share_num in result.allocated
+             }
+        )
+
+    def add_lease(self, storage_index, renew_secret, cancel_secret):
+        return self._client.add_lease(
+            storage_index, renew_secret, cancel_secret
+        )
+
+    @inlineCallbacks
+    def get_buckets(self, storage_index):
+        share_numbers = yield self._client.immutable_list_shares(
+            storage_index
+        )
+        returnValue({
+            share_num: _ClientV2BucketReader(
+                self._client, storage_index, share_num
+            )
+            for share_num in share_numbers
+        })
+
+    @inlineCallbacks
+    def slot_readv(self, storage_index, shares, readv):
+        reads = {}
+        for share_number in shares:
+            share_reads = reads[share_number] = []
+            for (offset, length) in readv:
+                d = self._client.mutable.read_share_chunk(
+                    storage_index, share_number, offset, length
+                )
+                share_reads.append(d)
+        result = {
+            share_number: [(yield d) for d in share_reads]
+            for (share_number, reads) in reads.items()
+        }
+        returnValue(result)
+
+    def slot_testv_and_readv_and_writev(
+            self,
+            storage_index,
+            secrets,
+            tw_vectors,
+            r_vector,
+    ):
+        we_secret, lr_secret, lc_secret = secrets
+        client_tw_vectors = {}
+        for share_num, (test_vector, data_vector, new_length) in tw_vectors.items():
+            assert new_length is not None, "Protocol in theory supports it, actual code seems not to"
```

Comment:
Which code doesn't support it?

Reply:
Actually I'm wrong, found code that does use it.